File: vanguard_system/hive_backend/models/telemetry_model.py
"""
Telemetry Data Model
====================
Pydantic model representing the drone telemetry payload.
All fields are typed and documented for clarity.
"""

from pydantic import BaseModel, Field
from datetime import datetime


class TelemetryData(BaseModel):
    """Complete telemetry snapshot from the drone."""

    drone_id: str = Field(
        ...,
        description="Unique identifier for the drone unit",
        examples=["VANGUARD-01"],
    )
    latitude: float = Field(
        ...,
        description="Current latitude in decimal degrees",
        ge=-90,
        le=90,
    )
    longitude: float = Field(
        ...,
        description="Current longitude in decimal degrees",
        ge=-180,
        le=180,
    )
    altitude: float = Field(
        ...,
        description="Altitude above ground level in metres",
        ge=0,
    )
    speed: float = Field(
        ...,
        description="Ground speed in m/s",
        ge=0,
    )
    battery: int = Field(
        ...,
        description="Battery level percentage",
        ge=0,
        le=100,
    )
    mode: str = Field(
        ...,
        description="Current flight mode (e.g. AUTO, LOITER, RTL)",
    )
    current_waypoint: int = Field(
        ...,
        description="Index of the waypoint the drone is currently heading to",
        ge=0,
    )
    total_waypoints: int = Field(
        ...,
        description="Total number of waypoints in the active mission",
        ge=0,
    )
    signal_strength: int = Field(
        ...,
        description="Communication link signal strength percentage",
        ge=0,
        le=100,
    )
    water_depth: float = Field(
        ...,
        description="Estimated water depth at the drone's current position in metres",
        ge=0,
    )
    timestamp: datetime = Field(
        ...,
        description="UTC timestamp of this telemetry snapshot",
    )

File: vanguard_system/mobile_app/jarvis_app/lib/models/telemetry_model.dart
/// Telemetry Data Model
/// ====================
/// Dart representation of the telemetry payload from the Hive API.
library;

class TelemetryData {
  final String droneId;
  final double latitude;
  final double longitude;
  final double altitude;
  final double speed;
  final int battery;
  final String mode;
  final int currentWaypoint;
  final int totalWaypoints;
  final int signalStrength;
  final double waterDepth;
  final DateTime timestamp;

  TelemetryData({
    required this.droneId,
    required this.latitude,
    required this.longitude,
    required this.altitude,
    required this.speed,
    required this.battery,
    required this.mode,
    required this.currentWaypoint,
    required this.totalWaypoints,
    required this.signalStrength,
    required this.waterDepth,
    required this.timestamp,
  });

  /// Parse from JSON map returned by the API.
  factory TelemetryData.fromJson(Map<String, dynamic> json) {
    return TelemetryData(
      droneId: json['drone_id'] as String? ?? 'UNKNOWN',
      latitude: (json['latitude'] as num?)?.toDouble() ?? 0.0,
      longitude: (json['longitude'] as num?)?.toDouble() ?? 0.0,
      altitude: (json['altitude'] as num?)?.toDouble() ?? 0.0,
      speed: (json['speed'] as num?)?.toDouble() ?? 0.0,
      battery: (json['battery'] as num?)?.toInt() ?? 0,
      mode: json['mode'] as String? ?? 'UNKNOWN',
      currentWaypoint: (json['current_waypoint'] as num?)?.toInt() ?? 0,
      totalWaypoints: (json['total_waypoints'] as num?)?.toInt() ?? 0,
      signalStrength: (json['signal_strength'] as num?)?.toInt() ?? 0,
      waterDepth: (json['water_depth'] as num?)?.toDouble() ?? 0.0,
      timestamp: DateTime.tryParse(json['timestamp'] as String? ?? '') ??
          DateTime.now(),
    );
  }

  /// Default empty telemetry for initial state.
  factory TelemetryData.empty() {
    return TelemetryData(
      droneId: '---',
      latitude: 0.0,
      longitude: 0.0,
      altitude: 0.0,
      speed: 0.0,
      battery: 0,
      mode: 'OFFLINE',
      currentWaypoint: 0,
      totalWaypoints: 0,
      signalStrength: 0,
      waterDepth: 0.0,
      timestamp: DateTime.now(),
    );
  }

  /// Whether this telemetry has real data (not the empty placeholder).
  bool get isConnected => droneId != '---';
}
